src/edu/psu/sweng500/team8/gui/SudokuGUI.java

```diff
@@ -346,12 +382,162 @@ private void radHardActionPerformed(java.awt.event.ActionEvent evt) {// GEN-FIRS
+              return "Sudoku files";
+           }
+        });
+		
+		fc.setAcceptAllFileFilterUsed(false);
 		int returnVal = fc.showOpenDialog(getComponent(0));
 
+		if(returnVal == JFileChooser.APPROVE_OPTION){
+		
+			SavePackage savePackage = null;
+	      try
+	      {
+	         FileInputStream fileIn = new FileInputStream(fc.getSelectedFile().getAbsolutePath());
+	         ObjectInputStream in = new ObjectInputStream(fileIn);
+	         savePackage = (SavePackage) in.readObject();
+	         in.close();
+	         fileIn.close();
+	         this.loadSession(savePackage);
+	      }catch(IOException i)
+	      {
+	         i.printStackTrace();
+	         return;
+	      }catch(ClassNotFoundException c)
+	      {
+	         System.out.println("Employee class not found");
+	         c.printStackTrace();
+	         return;
+	      }
+		}
+		
 	}// GEN-LAST:event_jButton11ActionPerformed
+	
 
+	
+    private void jButton10ActionPerformed(java.awt.event.ActionEvent evt) throws FileNotFoundException {
+		
+		
+		final JFileChooser fc = new JFileChooser();
+
+		fc.setFileFilter(new FileFilter()
+        {
+           @Override
+           public boolean accept(File file)
+           {
+              return (file.getName().toUpperCase().endsWith(".SUDOKU") || file.isDirectory());
+           }
+
+           @Override
+           public String getDescription()
+           {
+              return "Sudoku files";
+           }
+        });
+		
+		fc.setAcceptAllFileFilterUsed(false);
+		
+		int returnVal = fc.showSaveDialog(getComponent(0));;
+		
+		if(returnVal == JFileChooser.APPROVE_OPTION){
+			
+
+				File f = fc.getSelectedFile();
+				String path = f.getAbsolutePath();
+				
+				if(f.exists()){
+		            int result = JOptionPane.showConfirmDialog(this,"The file exists, overwrite?","Existing file",JOptionPane.YES_NO_CANCEL_OPTION);
+		            switch(result){
+		                case JOptionPane.YES_OPTION:
+		                    fc.approveSelection();
+		                   // savePuzzle(path);
+		                    return;
+		                case JOptionPane.NO_OPTION:
+		                    return;
+		                case JOptionPane.CLOSED_OPTION:
+		                    return;
+		                case JOptionPane.CANCEL_OPTION:
+		                    fc.cancelSelection();
+		                    return;
+		            }
+		        }
+		        fc.approveSelection();
+		        savePuzzle(path);
+		        
+		        gameBoard.remarkGivenCells(this.gameSession.getGameBoard().getCellGrid());
+		      }
+			}
+			
+    private void savePuzzle(String path){
+		try
+	      {
+		
+			if(!path.toLowerCase().endsWith(".sudoku")){
+				path = path + ".sudoku";
+			}
+			
+			SavePackage savePackage = new SavePackage();
+			savePackage.setCellGrid(this.gameSession.getGameBoard().getCellGrid());
+			savePackage.setPuzzle(this.gameSession.getGameBoard().getCurrentPuzzle());
+			savePackage.setControlGrid(this.gameBoard.getControlGrid());
```

Comment:
I mean this.gameSession.getGameBoard(). It is a coreDataStructures.Board which does not have any Swing components.